tests/unit/test_db.py

```diff
@@ -0,0 +1,141 @@
+
+        self.assertIsInstance(cursor, db.Cursor)
+
+    def test_gets_cursor_from_connection_with_explicit_session(self):
+        connection = db.Connection(
+            host='localhost', session=MagicMock(spec=httpx.Client))
+        connection.session.is_closed = False
+
+        cursor = connection.cursor()
+
+        self.assertIsInstance(cursor, db.Cursor)
+
+    def test_renews_session_if_closed_when_getting_cursor(self):
+        connection = db.Connection(host='localhost')
+        connection.cursor()
+        session1 = connection.session
+
+        session1.close()
+        connection.cursor()
+        session2 = connection.session
+
+        self.assertIsNot(session1, session2)
+
+    def test_starts_not_closed(self):
+        connection = db.Connection(
+            host='localhost', session=MagicMock(spec=httpx.Client))
+        cursor = connection.cursor()
+
+        self.assertFalse(connection.closed)
+        self.assertFalse(cursor.closed)
+
+    def test_closes_connection(self):
+        connection = db.Connection(
+            host='localhost', session=MagicMock(spec=httpx.Client))
+        cursor = connection.cursor()
+
+        connection.close()
+
+        self.assertTrue(connection.closed)
+        self.assertTrue(cursor.closed)
+
+    def test_closes_connection_even_if_cursor_already_closed(self):
+        connection = db.Connection(
+            host='localhost', session=MagicMock(spec=httpx.Client))
+        cursor = connection.cursor()
+        cursor.close()
+
+        connection.close()
+
+        self.assertTrue(connection.closed)
+        self.assertTrue(cursor.closed)
+
+    def test_closes_underlying_session_as_well(self):
+        connection = db.Connection(
+            host='localhost', session=MagicMock(spec=httpx.Client))
+        # Just to simulate an implicitly created session.
+        connection.is_session_external = False
+
+        connection.close()
+
+        self.assertTrue(connection.session.close.called)
+
+    def test_cant_close_connection_twice(self):
+        connection = db.Connection(
+            host='localhost', session=MagicMock(spec=httpx.Client))
+
+        connection.close()
+
+        with self.assertRaises(exceptions.Error):
+            connection.close()
+
+    def test_commits_nothing(self):
+        """
+        This is just a sanity test, to make sure we follow the expected
+        interface.
+        """
+        connection = db.Connection()
+
+        connection.commit()
+
+    def test_executes_a_statement(self):
+        connection = db.Connection(
+            host='localhost', session=MagicMock(spec=httpx.Client))
+        connection.session.is_closed = False
+        response = connection.session.post.return_value
+        response.json.return_value = {
+            'numServersResponded': 1,
+            'numServersQueried': 1,
+        }
+        response.status_code = 200
+
+        cursor = connection.execute('some statement')
```

Comment:
what exactly is this testing against? since the session is a MagicMock.

Reply:
This tests whether the Connection object is able to, generically, run a statement by delegating the request to the session object. Let me know if you'd prefer I add a docstring to explain the test a bit better.

Reply:
docstring added. Let me know what you think.

To be honest, I'm not completely satisfied with how the dependencies are played out there, so this test looks a bit weird - it's mocking something that is not a direct neighbor object to Connection. But I really prefer to first have the project covered with tests first, before I can take a stab at trying to improve the structure.

Reply:
Sounds good. Thanks. I saw similar test on other python dbapi projects as well just curious

Reply:
Got it